packages/navie/src/lib/apply-context.ts

```diff
@@ -0,0 +1,144 @@
+  let availableItemCount = availableContext.length;
+  while (availableItemCount > 0) {
+    const itemType = index % roundSize;
+    index += 1;
+
+    let contextItem: ContextV2.ContextItem | undefined;
+    const itemIndex = availableContext.findIndex(
+      (item) => item?.type === itemTypeByIndex[itemType]
+    );
+    if (itemIndex !== -1) {
+      availableItemCount -= 1;
+      contextItem = availableContext[itemIndex];
+      availableContext[itemIndex] = undefined;
+    }
+
+    if (contextItem) selectedContext.push(contextItem);
+  }
+
+  let charsRemaining = characterLimit;
+  log(`Remaining characters before context: ${charsRemaining}`);
+  const appliedContext = new Array<ContextV2.ContextItem>();
+
+  const itemDescription = (contextItem: ContextV2.ContextItem): string =>
+    [contextItem.type, contextItem.location].filter(Boolean).join(' ');
+
+  const addContextItem = (contextItem: ContextV2.ContextItem): ContextItemStatus => {
+    // Don't consume too much of the character limit on a single item.
+    if (contextItem.content.length > maxContentLength) {
+      if (verbose) {
+        log(
+          `Skipping context item ${itemDescription(contextItem)} due to size. ${
+            contextItem.content.length
+          } > ${maxContentLength}`
+        );
+      }
+      return ContextItemStatus.ITEM_TOO_BIG;
+    }
+
+    if (
+      contextItem.type === ContextV2.ContextItemType.SequenceDiagram &&
+      appliedContext[appliedContext.length - 1]?.type === ContextV2.ContextItemType.SequenceDiagram
+    ) {
+      if (verbose)
+        log(
+          `Skipping context item ${itemDescription(
+            contextItem
+          )} because the previous context item is also a sequence diagram.`
+        );
+      return ContextItemStatus.SIMILAR_CONTENT;
+    }
+
+    if (verbose)
+      log(`Adding context item ${itemDescription(contextItem)} (${contextItem.content.length})`);
+
+    charsRemaining -= contextItem.content.length;
+    appliedContext.push(contextItem);
+
+    return ContextItemStatus.OK;
+  };
+
+  let hasSequenceDiagramBeenAdded = false;
+  const onlySequenceDiagramsRemaining = (i: number) =>
+    selectedContext
+      .slice(i)
+      .every((item) => item.type === ContextV2.ContextItemType.SequenceDiagram);
+
+  for (let i = 0; i < selectedContext.length; i += 1) {
+    const contextItem = selectedContext[i];
+    const status = addContextItem(contextItem);
+    if (
+      status === ContextItemStatus.OK &&
+      contextItem.type === ContextV2.ContextItemType.SequenceDiagram
+    ) {
+      hasSequenceDiagramBeenAdded = true;
+    }
+
+    if (verbose) log(`Context item ${itemDescription(contextItem)} status: ${status}`);
+
+    if (hasSequenceDiagramBeenAdded && onlySequenceDiagramsRemaining(i)) {
+      log(
+        `Only sequence diagrams remain in the context, and at least one sequence diagram has already been added. No further context will be added.`
+      );
+      break;
+    }
```

Comment:
Should this occur before `addContextItem` is called, or should it be `onlySequenceDiagramsRemaining(i + 1)`?

Otherwise this will consider the current context item, which could then be followed by only sequence diagrams. Maybe this is okay but the log message makes it look like this isn't intended.